### ISSUE TITLE
Add support for streaming and ultrasound data

### ISSUE THREAD

**Issue by KumoLiu** [Project-MONAI/MONAI#8172]

**Comment by mingxin-zheng** [comment]
Hi @KumoLiu , thanks for the proposal. I think this task lacks some definition right now, as post-processed ultrasound data in DICOM format is not that special. If we are trying to work with raw ultrasound data (e.g. phase array or curvilinear array), then it is a very different story.

Maybe we need some ultrasound experts to suggest what the community may need. cc @ericspod @aylward @Nic-Ma Thanks!

**Comment by ericspod** [comment]
Hi @mingxin-zheng I think we do need some specialist opinions here for sure! I can imagine needing support for loading/saving new data formats but also transforms for working specifically with US data. These would cover common processing operations but also generating realistic noise or other signal degradation specific to US for data augmentation.

**Comment by mingxin-zheng** [comment]
Hi @ericspod , it's a good point to bring the up the noise augmentation! As far as I know there are at least two special kinds there: noise introduced by speckle and imaging-depth.

On the other hand, what specific data formats have you seen with ultrasound systems besides DICOM?

**Comment by ericspod** [comment]
I'm not well versed in what common practices are with US so can't really comment on what formats there would be. We will discuss with colleagues involved with US research and see what requirements they come up with.

**Comment by jamesobutler** [comment]
Coming from an Ultrasound pre-clinical imaging group we use [PlusToolkit](https://plustoolkit.github.io/) (main source repo: https://github.com/PlusToolkit/PlusLib) along with [3D Slicer](https://www.slicer.org/) (main source repo: https://github.com/Slicer/Slicer) to collect ultrasound data from various ultrasound hardware and save to [MetaIO](https://github.com/Kitware/MetaIO) and [NRRD](https://teem.sourceforge.net/nrrd/) formats. In this case we are using PlusToolKit as an open-source solution for ultrasound streaming and recording. PlusToolKit’s PlusServer is the primary tool, but there are some other ultrasound processing tools (see http://perk-software.cs.queensu.ca/plus/doc/nightly/user/Tools.html).

The NA-MIC project weeks have also had ultrasound AI based projects such as https://projectweek.na-mic.org/PW39_2023_Montreal/Projects/LiveTrackedUltrasoundProcessingWithPytorch/.

**Comment by mingxin-zheng** [comment]
Thanks @jamesobutler for the great inputs!

Today I met @bcjiang who also works in this domain. Wondering if you can share some experiences too

**Comment by bcjiang** [comment]
Hi @mingxin-zheng, we typically obtain the ultrasound raw channel data from the Verasonix machines. It has a MATLAB interface and the channel data is saved as a multi-dimensional array with 2D Frame index, A-line index and Signal data in each dimension. Each element in Signal data is in 14-bit signed encoding (-8192 to 8191). We can save these data to any file type and load them into python. 
I think prior DL works leveraging the raw channel US data were mostly aiming at beamforming/imaging quality improvement, so data can be easily acquired by just scanning different subjects continuously. I didn't see much augmentation was applied in these works, but I'm also interested to learn some other augmentation methods for raw data in classification/segmentation scenarios.

**Comment by mingxin-zheng** [comment]
Thanks @bcjiang ! The raw data is interesting. Does the .mat file also include probe apeture information for scan conversion too, or that's a seperate piece of data?

**Comment by bcjiang** [comment]
Hi @mingxin-zheng, yes there is also a header structure array in MATLAB workspace that we can export, which is created by a self-defined setup script before running the Verasonix data acquisition. The header should contain all probe information and we can choose which part to export and specify the export file format.

**Comment by mingxin-zheng** [comment]
Thanks @bcjiang for the explanation ! 

Now I'm just imagining if we need to implement such conversion for ultrasound data and here could be one thing for MONAI `MetaTensor` to improve:
Typically, the `MetaTensor` class contains an affine matrix to represent the spatial transform applied to an image (2D/3D). But ultrasound's scan conversion, cann't be represented by one single affine operation. Each data line has its own spatial transform (e.g. angle to steer), when it needs to be interpolated to a orthogonal raster.

Here is an illustration of such process. It needs to be noted that **B** is the RF raw data and **A** is what we see after scan-conversion and interpolation.

![Back-scan-conversion-a-scheme-of-original-sector-B-mode-image-b-same-after](https://github.com/user-attachments/assets/f7bf5955-b40e-40fb-b35f-c23c98a67484)

**Comment by jamesobutler** [comment]
Is MONAI considering tasks associated with RF to scan converted images? Or taking Ultrasound scan converted images from ultrasound hardware devices and then processing them for various labeling tasks? Often times Ultrasound hardware provides both scan converted and the raw RF data.

**Comment by mingxin-zheng** [comment]
I think this is still open to discussion here @jamesobutler . It is still a research topic how much RF data can provide beyond scan-converted images. I personally think MONAI can consider features to accelerate such explorations. But what features are needed remains undefined at the moment.

**Comment by jamesobutler** [comment]
Ok. Yea ultrasound scan conversion can be different depending on the type of probe. A linear array, curved array, phased array, matrix array or single element or annular array types can produce different looking scan converted images. We use a high frequency linear array now for most of our work with various techniques such as spatial compounding multiple angles to form a single processed image or using multiple focal zones which acquire multiple frames that process into a single image.

With ultrasound when trying to get good input images for MONAI models we are often managing how imaging depth is impacting attenuation, usage of TGC impacting signal intensity along that depth, shadowing caused by interfaces such as bone or air and speckle adding various intensity values across the image.

**Comment by aylward** [comment]
I vote for supporting both RF and B-mode (including B-mode video).

Regarding special processing for ultrasound...in addition to creating B-Mode from RF (transducer lines), it is often useful to pre-process the data from B-mode images back into transducer lines (not RF, but the reconstructed signal along a transducer line) when a convex probe is used.   There are a few reasons for this.  (1) Most b-mode images from convex probes are nearly 50% blank / irrelevant.  (2) Image information/detail near a convex transducer is spatially more dense than the information density deeper into the patient (e.g., near the bottom of the image).  (3) The impact of signal attenuation (e.g., shadowing) follows transducer lines whereas in b-mode they will spread out to the side depending on where they fall in the image.  All of these features means that when rectangular convolutional kernels are used, it is often "better" to apply them in transducer-line space rather than in b-mode images - transducer-line space does not have blank space, has more uniform information density, and attenuation/artifacts have vertical (rectangular, not angular) impact.

I used these transforms for several projects with DARPA and DoD - with using transducer-line space greatly reducing the amount of training data needed compared to doing things in b-mode space.

I've also done a fair amount of Quantitative US (QUS) work in which the RF is process directly, instead of being reconstructed into b-mode images.   QUS is going to be the future of ultrasound, IMO :)

**Comment by mingxin-zheng** [comment]
I totally agree with @aylward 's comment regarding processing from B-mode images back into transducer line because that's exactly what I heard a couple of years ago. Moreover, those works simulated (1) TGC gain level change (2) imaging depth change for curvilinear transducer for image augmentation. I remembered those augmentation would make the DL algorithm more robust to transducer settings when it gets to real-time deployment.

**Comment by mingxin-zheng** [comment]
Add a reference to implementations:
https://github.com/SlicerIGT/aigt/issues/50

**Comment by mingxin-zheng** [comment]
In the weekly MONAI meeting, tracked-ultrasound usage is mentioned, which may relate to the point transform feature developed in MONAI 1.4.
In this setting, monai transform can be used to compute the position of ultrasound image in 3D space. This typically requires (1) tracker positions from external device, e.g. optical reflective markers; (2) the relative position between the tracker and the ultrasound probe. With point transform, we can have both image and tracker position to be processed by a single MONAI compose.

**Comment by aylward** [comment]
@jamesobutler and others.  We're thinking of creating a MONAI working group on ultrasound.  Please send me an email if you're interested.   saylward at nvidia dot com

**Comment by aylward** [comment]
Sorry for not posting sooner, but the Ultrasound Working Group was approved by the board several months ago, and it is going extremely well!

For more info

* The email list:
https://groups.google.com/g/monai-wg-ultrasound/

* Meeting:
If you join the email list, you will be automatically added to the calendar invite for the monthly meeting:
11am Eastern Time zone USA 
Monthly: Third Wednesday
Meeting link: https://teams.microsoft.com/l/meetup-join/19%3ameeting_MzUxMDE1ODYtYWYzZC00ODhmLWEzMmItMzdkMzNkYTdjOWNj%40thread.v2/0?context=%7b%22Tid%22%3a%2243083d15-7273-40c1-b7db-39efd9ccc17a%22%2c%22Oid%22%3a%227f517236-a5e1-4d2f-a5f1-2ff018d08bcf%22%7d
Meeting ID: 231 541 490 314 1
Passcode: fc6zZ9Y4

* Slack:
Search for channels beginning with WG-Ultrasound:
https://join.slack.com/t/projectmonai/shared_invite/zt-3hucgm02q-i8Bn9XofDZs2UGOH4jUl4w
